src/tribler/core/components/user_activity/user_activity_component.py

```diff
@@ -0,0 +1,110 @@
+        self.max_query_history = self.session.config.user_activity.max_query_history
+
+        # Wait for dependencies
+        await self.require_component(ContentDiscoveryComponent)  # remote_query_results notification
+        await self.require_component(LibtorrentComponent)  # torrent_finished notification
+        database_component = await self.require_component(DatabaseComponent)  # local_query_results notification
+        torrent_checker_component = await self.require_component(TorrentCheckerComponent)
+
+        self.database_manager: UserActivityLayer = database_component.db.user_activity_layer
+        self.torrent_checker: TorrentChecker = torrent_checker_component.torrent_checker
+
+        # Hook events
+        self.session.notifier.add_observer(notifications.torrent_finished, self.on_torrent_finished)
+        self.session.notifier.add_observer(notifications.remote_query_results, self.on_query_results)
+        self.session.notifier.add_observer(notifications.local_query_results, self.on_query_results)
+        self.task_manager.register_task("Check preferable", self.check_preferable,
+                                        interval=self.session.config.user_activity.health_check_interval)
+
+    def on_query_results(self, data: dict) -> None:
+        """
+        Start tracking a query and its results.
+
+        If any of the results get downloaded, we store the query (see ``on_torrent_finished``).
+        """
+        query = data.get("query")
+        if query is None:
+            return
+
+        results = {tmd.infohash for tmd in data["results"]}
+        for infohash in results:
+            self.infohash_to_queries[infohash].append(query)
+        self.queries[query] = results | self.queries.get(query, set())
+
+        if len(self.queries) > self.max_query_history:
+            query, results = self.queries.popitem(False)
+            for infohash in results:
+                self.infohash_to_queries[infohash].remove(query)
+                if not self.infohash_to_queries[infohash]:
+                    self.infohash_to_queries.pop(infohash)
+
+
+    @staticmethod
+    async def _fut_to_task(future: Future):
+        await future
+
+    def on_torrent_finished(self, infohash: str, name: str, hidden: bool) -> None:
+        """
+        When a torrent finishes, check if we were tracking the infohash. If so, store the query and its result.
+        """
+        b_infohash = InfoHash(unhexlify(infohash))
+        queries = self.infohash_to_queries[b_infohash]
+        for query in queries:
+            losing_infohashes = self.queries[query] - {b_infohash}
+            fut = get_running_loop().run_in_executor(None,
+                                                     self.database_manager.store,
+                                                     query, b_infohash, losing_infohashes)
+            self.task_manager.register_task("Store query", functools.partial(UserActivityComponent._fut_to_task, fut))
```

Comment:
Two points:
1. This method seems overly complicated for its simple purpose — to call the `self.database_manager.store` function.
2. I'm uncertain if the logic will work correctly in cases where the user downloads two or more torrents from the query results.

Reply:
This is solving a real issue: it is related to #4320 and #7784.

IIRC, the observation that we should not block the main `asyncio` thread with db operations should be accredited to @egbertbouman but I cannot find his original comment for the life of me. Once I find it, I'll properly link it.

Reply:
I did manage to shrink this a little bit (see [conversation with @kozlovsky](https://github.com/Tribler/tribler/pull/7786#discussion_r1439313557)).